Allow Agent Reload with API - /v1/agent/reload

benjamin-lykins opened the pull request:
### Description
Provide an API endpoint to reload agent config without elevated permissions on the infrastructure running Nomad. 
- Provides ability to reload config remotely. 
- Leverage Nomad Token + ACL to reload. 

### Testing 

**Reloading log level without ACL**
```sh
curl -k  -X PUT \
  https://localhost:4646/v1/agent/reload
{"message":"agent configuration reloaded"}
```

**Attempting without token (ACL Enabled):** 
```sh
curl -k  -X PUT \  
https://localhost:4646/v1/agent/reload
Permission denied
```

**Attempting with token (ACL Enabled; `Agent = Write`):** 

```sh
curl -k -X PUT \                                                    
  --header "X-Nomad-Token: 79d77f6a-e168-3bbb-8fa7-e668b2b09346" \
  https://localhost:4646/v1/agent/reload 
{"message":"agent configuration reloaded"}
```

**Attempting with token (ACL Enabled; `Agent = Read`):** 

```sh
curl -k -X PUT \
  --header "X-Nomad-Token: 2cde0074-d0ac-53e1-a6dc-3f416a8e1737" \
  https://localhost:4646/v1/agent/reload
Permission denied
```

**Invalid Config File:**

```sh
curl -k -X PUT \
  --header "X-Nomad-Token: 79d77f6a-e168-3bbb-8fa7-e668b2b09346" \
  https://localhost:4646/v1/agent/reload
Error loading /Users/benjamin.lykins/git/nomad/nomad/bin/config/general.hcl: failed to decode HCL file /Users/benjamin.lykins/git/nomad/nomad/bin/config/general.hcl: At 3:16: literal not terminated
```

### Links
Fixes: #24048

### Contributor Checklist
- [x] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [X] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [ ] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository. 


<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [ ] If a change needs to be reverted, we will roll out an update to the code within 7 days.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

benjamin-lykins committed:
docs: update nomad.service to point to the current site

benjamin-lykins committed:
Merge branch 'hashicorp:main' into main

benjamin-lykins committed:
Merge branch 'hashicorp:main' into main

benjamin-lykins committed:
Merge branch 'hashicorp:main' into main

benjamin-lykins committed:
api: add reload endpoint to agent for configuration reload

benjamin-lykins committed:
api: add simple returned message to reload endpoint

benjamin-lykins committed:
test: add tests for agent reload endpoint

benjamin-lykins committed:
api: update notes

benjamin-lykins committed:
api: update return message

benjamin-lykins committed:
api: add logging for agent reload process

benjamin-lykins committed:
chore: cleaning up my comments and notes

benjamin-lykins commented on command/agent/agent_endpoint.go:
```diff
@@ -563,6 +564,55 @@ func (s *HTTPServer) listServers(resp http.ResponseWriter, req *http.Request) (i
 
+type reloadResponse struct {
+	Message string `json:"message"`
+}
```
simple response back for now (essentially if you don't get an error you will get this back). changes could be made to the reload method to return if it did make changes, but there are not many right now.

benjamin-lykins committed:
api: cleanup

tgross reviewed (changes requested):
@benjamin-lykins there's a failing test not touched here but clearly related to this work. Let's get that resolved before we review:

```
=== FAIL: command/agent TestServer_Reload_TLS_Certificate (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x30fefa3]

goroutine 87962 [running]:
testing.tRunner.func1.2({0x3898300, 0x69a5e70})
	/opt/hostedtoolcache/go/1.25.3/x64/src/testing/testing.go:1872 +0x237
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.25.3/x64/src/testing/testing.go:1875 +0x35b
panic({0x3898300?, 0x69a5e70?})
	/opt/hostedtoolcache/go/1.25.3/x64/src/runtime/panic.go:783 +0x132
github.com/hashicorp/nomad/command/agent.(*Agent).Reload(0xc00c83c700, 0xc00d476d08)
	/home/runner/work/nomad/nomad/command/agent/agent.go:1533 +0x63
github.com/hashicorp/nomad/command/agent.TestServer_Reload_TLS_Certificate(0xc00a9f8540)
	/home/runner/work/nomad/nomad/command/agent/agent_test.go:1197 +0x1d7
testing.tRunner(0xc00a9f8540, 0x42c8720)
	/opt/hostedtoolcache/go/1.25.3/x64/src/testing/testing.go:1934 +0xea
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.25.3/x64/src/testing/testing.go:1997 +0x465

DONE 5542 tests, 12 skipped, 4 failures in 182.433s
```

benjamin-lykins committed:
api: remove debug log from agent reload method, fix failing test

benjamin-lykins committed:
api: removing response validation in testing, check fo error and not nil only

tgross commented on command/agent/agent_endpoint.go:
```diff
@@ -563,6 +564,52 @@ func (s *HTTPServer) listServers(resp http.ResponseWriter, req *http.Request) (i
+
+	newConf.Files = append([]string(nil), currConf.Files...)
+
+	if err := s.agent.Reload(newConf); err != nil {
```
If you take a look at the signal handler for reloading, you'll see that this step only reloads the agent configuration and won't push any changes to the server, client, or HTTP server configuration. We should at least have parity with the signal handler here, which calls `(*Command) handleReload`. We may need to figure out a way to extract that function so that it's reachable here.

tgross reviewed (changes requested):
This is a good start @benjamin-lykins!

In addition to the comments I've left, make sure you've added a changelog entry file. You can do that via `make cl`. Once this is getting closer to being ready, we'll need a docs PR to go with this, which is now going to be over in the [`web-unified-docs`](https://github.com/hashicorp/web-unified-docs) repo.

benjamin-lykins committed:
api: update Reload method to accept AgentReloadOpts for future configuration options

benjamin-lykins committed:
api: update tests to use shoenig/test instead of testify, consolidate tests into one

benjamin-lykins committed:
api: testing clean up

benjamin-lykins committed:
config: store -config paths on startup

benjamin-lykins committed:
api: fix reload api to use stored -config values when checking for new config values/files

benjamin-lykins committed:
tests: removed old commented code

benjamin-lykins committed:
api: move handleReload for parity between API and SIGHUP (at least i hope so)

tgross reviewed (changes requested):
I think I see where you're going with this, but there's an architectural issue. See my comment for more details.

benjamin-lykins committed:
api: revert functionality, add parity between api and SIGHUP

benjamin-lykins committed:
api: simplify comments in Agent struct and remove unnecessary whitespace

benjamin-lykins committed:
removing unused test

benjamin-lykins committed:
clean up

benjamin-lykins committed:
Merge branch 'main' into feat/config-reload-api

benjamin-lykins committed:
running gofmt

mismithhisler commented on command/agent/agent.go:
```diff
@@ -1607,6 +1611,18 @@ func (a *Agent) ShouldReload(newConfig *Config) (agent, http bool) {
+// receiving a SIGHUP signal. The reload logic lives in Command.handleReload.
+func (a *Agent) ConfigReload() error {
+	a.configLock.Lock()
+	reloader := a.configReloader
```
Is it necessary to lock this and make this copy here? This is set on the agent before starting the HTTP servers, so nobody could call it before the agent's configReloader is set.

benjamin-lykins committed:
removing lock/unlock

benjamin-lykins committed:
removing unnecessary variable

benjamin-lykins committed:
fix: now picking up and logging newly found config files

benjamin-lykins commented:
@tgross ended up reworking based on your last review. 

=> Test

1. Setup

command :
`.nomad agent -config=pkg/darwin_arm64/config/server -config=pkg/darwin_arm64/config/global.hcl`

This set up an entire directory: `server` and a specific file : `global.hcl`

```sh
➜  nomad git:(feat/config-reload-api) curl -s http://localhost:4646/v1/agent/self | jq .config.Files
[
  "pkg/darwin_arm64/config/server/server.hcl",
  "pkg/darwin_arm64/config/global.hcl"
```

Current log level
```sh
➜  nomad git:(feat/config-reload-api) curl -s http://localhost:4646/v1/agent/self | jq .config.LogLevel
"INFO"
```

2. Config change

Update the server directory to be multiple files. 

- move `data_dir` config into its own file `data_dir.hcl`
  - this doesn't do much but confirms it picks up the new file once I reload. 
- added `log_level.hcl`. 
  - Change log level from `INFO` to `DEBUG`. 

3. Reload via API

Ran the following:

```sh
curl -k  -X PUT \
  https://localhost:4646/v1/agent/reload
```

Result:

```sh
➜  nomad git:(feat/config-reload-api) curl -k -X PUT http://localhost:4646/v1/agent/reload 
{"message":"agent configuration reloaded"}%     
```

Logs:

```sh
==> Reloading configuration...
==> WARNING: mTLS is not configured - Nomad is not secure without mTLS!
==> WARNING: Bootstrap mode enabled! Potentially unsafe operation.
    2026-05-14T23:36:27.200-0400 [INFO]  agent: detected new config files before reload: files="pkg/darwin_arm64/config/server/data_dir.hcl, pkg/darwin_arm64/config/server/log_level.hcl"
    2026-05-14T23:36:27.200-0400 [DEBUG] agent: starting reload of server config
    2026-05-14T23:36:27.200-0400 [DEBUG] http: request complete: method=PUT path=/v1/agent/reload duration=225.535167ms
```

Log Level Update:

```sh
➜  nomad git:(feat/config-reload-api) curl -s http://localhost:4646/v1/agent/self | jq .config.LogLevel
"DEBUG"
```          

New config files are getting picked up: 

```sh
➜  nomad git:(feat/config-reload-api) ✗ curl -s http://localhost:4646/v1/agent/self | jq .config.Files
[
  "pkg/darwin_arm64/config/server/data_dir.hcl",
  "pkg/darwin_arm64/config/server/log_level.hcl",
  "pkg/darwin_arm64/config/server/server.hcl",
  "pkg/darwin_arm64/config/global.hcl"
]
```

FYI: I set this up to log new files found when reloading. I can drop to debug from info, but its not going to be too common I think. 

```sh
	newFiles := checkNewConfigFiles(oldConf.Files, newConf.Files)
	if len(newFiles) > 0 {
		c.agent.logger.Info("detected new config files before reload", "files", strings.Join(newFiles, ", "))
	}
```

tgross commented on api/agent.go:
```diff
@@ -219,6 +219,15 @@ func (a *Agent) ListKeys() (*KeyringResponse, error) {
+type AgentReloadOpts struct{}
+
+// Reload requests the agent to reload its configuration.
+func (a *Agent) Reload(_ *AgentReloadOpts) error {
```
This needs a `WriteOptions` parameter as well so that any Go API consumers can pass their auth token.

benjamin-lykins replied:
added WriteOptions in here::: 

```go
type AgentReloadOpts struct {
	WriteOptions
}

// Reload requests the agent to reload its configuration.
func (a *Agent) Reload(q *AgentReloadOpts) error {
	var writeOpts *WriteOptions
	if q != nil {
		writeOpts = &q.WriteOptions
	}
	_, err := a.client.put("/v1/agent/reload", nil, nil, writeOpts)
	return err
}
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			must.NotNil(t, obj)
+		})
+	})
+}
```
When you moved the invalid method test into the `TestHTTP_AgentReload_ACL` this test is no longer telling us anything new. We can drop it.

```suggestion
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+
+		// Make the HTTP request
+		req, err := http.NewRequest(http.MethodPut, "/v1/agent/reload", nil)
+		must.Nil(t, err)
```
This gives a nicer assertion message:

```suggestion
		must.NoError(t, err)
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+		// Try request with an invalid method
+		{
+			req, err := http.NewRequest(http.MethodGet, "/v1/agent/reload", nil)
+			must.Nil(t, err)
```
```suggestion
			must.NoError(t, err)
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			respW := httptest.NewRecorder()
+			_, err = s.Server.AgentReloadRequest(respW, req)
+			must.NotNil(t, err)
+			must.Eq(t, ErrInvalidMethod, err.Error())
```
```suggestion
			must.EqError(t, err, ErrInvalidMethod)
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			respW := httptest.NewRecorder()
+			_, err := s.Server.AgentReloadRequest(respW, req)
+			must.NotNil(t, err)
+			must.Eq(t, structs.ErrPermissionDenied.Error(), err.Error())
```
```suggestion
			must.EqError(t, err, structs.ErrPermissionDenied.Error())
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			token := mock.CreatePolicyAndToken(t, state, 1005, "invalid", mock.NodePolicy(acl.PolicyWrite))
+			setToken(req, token)
+			_, err := s.Server.AgentReloadRequest(respW, req)
+			must.NotNil(t, err)
```
```suggestion
			must.EqError(t, err, structs.ErrPermissionDenied.Error())
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			setToken(req, token)
+			_, err := s.Server.AgentReloadRequest(respW, req)
+			must.NotNil(t, err)
+			must.Eq(t, structs.ErrPermissionDenied.Error(), err.Error())
```
```suggestion
			must.EqError(t, err, structs.ErrPermissionDenied.Error())
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			token := mock.CreatePolicyAndToken(t, state, 1007, "valid", mock.AgentPolicy(acl.PolicyWrite))
+			setToken(req, token)
+			obj, err := s.Server.AgentReloadRequest(respW, req)
+			must.Nil(t, err)
```
```suggestion
			must.NoError(t, err)
```

tgross commented on command/agent/agent_endpoint_test.go:
```diff
@@ -2277,3 +2277,78 @@ func TestHTTP_AgentSchedulerWorkerConfigRequest_Client(t *testing.T) {
+			setToken(req, token)
+			obj, err := s.Server.AgentReloadRequest(respW, req)
+			must.Nil(t, err)
+			must.NotNil(t, obj)
```
It's great we have all these test cases... but we're not asserting that it actually worked!

benjamin-lykins replied:
Made this test a bit more robust

```go
		// Try request with a valid write token
		// Update from trace to warn log levels
		// Warn makes a pretty noticeable change
		{
			respW := httptest.NewRecorder()
			token := mock.CreatePolicyAndToken(t, state, 1007, "valid", mock.AgentPolicy(acl.PolicyWrite))
			setToken(req, token)

			// Grab the initial log level.
			initalLogLevel := s.Agent.logger.GetLevel().String()
			reloadedLogLevel := "WARN"

			reloadCalls := 0
			s.Agent.configReloader = func() error {
				reloadCalls++
				newConfig := s.Agent.GetConfig().Copy()
				newConfig.LogLevel = reloadedLogLevel
				return s.Agent.Reload(newConfig)
			}

			obj, err := s.Server.AgentReloadRequest(respW, req)
			must.NoError(t, err)
			must.Nil(t, obj)                                            // this will need changed if the API ever returns something other than nil
			must.Eq(t, 1, reloadCalls)                                  // confirm that the reloader was called 
			must.NotEq(t, initalLogLevel, s.Agent.GetConfig().LogLevel) // confirm that the log level was actually changed from intial value
			must.Eq(t, reloadedLogLevel, s.Agent.GetConfig().LogLevel)  // confirm that the log level was updated in the agent's config
		}
```

tgross commented on command/agent/agent_endpoint.go:
```diff
@@ -563,6 +563,35 @@ func (s *HTTPServer) listServers(resp http.ResponseWriter, req *http.Request) (i
+		Message: "agent configuration reloaded",
+	}
+
+	return response, nil
```
This object doesn't really give us anything interesting to work with. I'd probably return `nil` here and let the 200 OK speak for itself, similar to how we've done similar requests where we're not hitting the state store.

benjamin-lykins replied:
yeah I had a thought to grab what was changing and return it back, but it might not be really necessary at this point.

tgross commented on command/agent/agent_endpoint.go:
```diff
@@ -563,6 +563,35 @@ func (s *HTTPServer) listServers(resp http.ResponseWriter, req *http.Request) (i
+	}
+
+	if err := s.agent.ConfigReload(); err != nil {
+		s.logger.Error("failed to reload agent configuration", "error", err, "path", "/v1/agent/reload", "method", req.Method)
```
Because the `handleReload` logs every error, whenever there's an error, we end up with a repeated log line:

> 2026-05-18T13:59:58.729-0400 [ERROR] agent: reloading client config failed: error="VerifyOutgoing set, and no CA certificate provided!"
> 2026-05-18T13:59:58.729-0400 [ERROR] http: failed to reload agent configuration: error="reloading client config failed: VerifyOutgoing set, and no CA certificate provided!" path=/v1/agent/reload method=PUT
> 2026-05-18T13:59:58.729-0400 [ERROR] http: request failed: method=PUT path=/v1/agent/reload error="reloading client config failed: VerifyOutgoing set, and no CA certificate provided!" code=500
> 2026-05-18T13:59:58.729-0400 [DEBUG] http: request complete: method=PUT path=/v1/agent/reload duration=23.269117ms

So I think we can drop this line.


```suggestion
```

benjamin-lykins replied:
removed

tgross commented on command/agent/agent.go:
```diff
@@ -135,6 +135,10 @@ type Agent struct {
 
 	inmemSink *metrics.InmemSink
 
+	// configReloader triggers a full agent configuration
+	// reload, equivalent to SIGHUP
+	configReloader func() error
```
Something jumps out at me when I'm looking at this bit of dependency injection, and the log outputs generated from `handleReload`'s calls to `c.Ui`. 

We e probably want to move `handleReload` out of `Command` entirely and into `Agent`. It clearly doesn't belong there and non-agent commands certainly don't want to handle a config reload. But that's refactoring for a later PR.

tgross commented on command/agent/command.go:
```diff
@@ -1249,9 +1262,41 @@ func (c *Command) handleReload() error {
+	c.agent.configLock.Lock()
+	c.agent.config.Files = append([]string(nil), newConf.Files...)
+	c.agent.config.ConfigPaths = append([]string(nil), newConf.ConfigPaths...)
+	c.agent.configLock.Unlock()
```
I would push this down into `Agent.Reload` and add a case for `Agent.ShouldReload` where if the configuration files are different, reload the agent. That'll prevent you from having to take the lock again or allocating these slices. I think you can avoid adding the extra `ConfigPaths` field as well in that case.

tgross reviewed (commented):
This is looking great @benjamin-lykins. I think there's one more minor architectural issue I've noted, and a couple of odds-and-ends and we should be able to ship this.

We'll need a changelog in this PR, and API docs in the docs repo as well.

benjamin-lykins committed:
Update command/agent/agent_endpoint_test.go

Co-authored-by: Tim Gross <tim@0x74696d.com>

benjamin-lykins committed:
adding write options.

benjamin-lykins committed:
removing response and log from agent_endpoint

benjamin-lykins committed:
first batch of test changes per review

benjamin-lykins committed:
tests : make the valid test more robust. change from trace to warn.

tgross commented:
@benjamin-lykins you're going to start seeing copyright header checks fail, because we've updated our `copywrite` config to match the actual policy for headers. Any file you touch should have the last date updated to 2026. Ex.

```go
// Copyright IBM Corp. 2015, 2025
// SPDX-License-Identifier: BUSL-1.1
```

should be

```go
// Copyright IBM Corp. 2015, 2026
// SPDX-License-Identifier: BUSL-1.1
```

benjamin-lykins committed:
update copywrite header

tgross commented on api/agent.go:
```diff
@@ -1,5 +1,5 @@
-// Copyright IBM Corp. 2015, 2025
-// SPDX-License-Identifier: MPL-2.0
+// Copyright IBM Corp. 2015, 2026
+// SPDX-License-Identifier: BUSL-1.1
```
Just the year needs to change. The `api` package is one of a few exceptions to the BUSL license.

```suggestion
// SPDX-License-Identifier: MPL-2.0
```

But looks like you've got a merge conflict in this file which I bet is from the header anyways

tgross commented on api/agent.go:
```diff
@@ -219,6 +219,20 @@ func (a *Agent) ListKeys() (*KeyringResponse, error) {
 
+type AgentReloadOpts struct {
+	WriteOptions
+}
```
We generally don't embed these in the API package. I don't actually have strong feelings about the approach one way or the other, but consistency is important.

```suggestion
type AgentReloadOpts struct {}
```

tgross commented on api/agent.go:
```diff
@@ -219,6 +219,20 @@ func (a *Agent) ListKeys() (*KeyringResponse, error) {
+	if q != nil {
+		writeOpts = &q.WriteOptions
+	}
+	_, err := a.client.put("/v1/agent/reload", nil, nil, writeOpts)
```
Splitting out the `WriteOptions`, which are safely handled if nil in `a.client.put`:

```suggestion
func (a *Agent) Reload(_ *AgentReloadOpts, q *WriteOptions) error {
	_, err := a.client.put("/v1/agent/reload", nil, nil, wo)
```

benjamin-lykins committed:
Apply suggestions from code review

Co-authored-by: Tim Gross <tim@0x74696d.com>

benjamin-lykins committed:
Fix agent reload function to use correct WriteOptions and update copyright year

benjamin-lykins committed:
fixing last copywrite failure

benjamin-lykins committed:
sigh.

benjamin-lykins committed:
updated per comments, and validated manual + go test.
- removed logic/lock from command.go
- shouldReload updated to check  for differences in config path/files

tgross reviewed (commented):
LGTM, @benjamin-lykins, except for one last item I forgot to check earlier: a changelog! If you run `make cl` in the root of the repo it'll generate the correctly-shaped file for this feature.

benjamin-lykins committed:
cl : adding notes

benjamin-lykins commented:
> LGTM, @benjamin-lykins, except for one last item I forgot to check earlier: a changelog! If you run `make cl` in the root of the repo it'll generate the correctly-shaped file for this feature.

My bad! It was sitting locally and I forgot to push it up.

tgross reviewed (approved):
LGTM!
